### PR TITLE
NEXT-12594 - Add using state machine guide

### PR DESCRIPTION
**Pull request by PaddyS**

### CODE REVIEW

**Review by shyim — APPROVED** [review]

File: guides/plugins/plugins/checkout/order/using-the-state-machine.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+- `cancel`: Will result in "cancelled"
+- `retour`: Will result in "returned"
+- `retour_partially`: Will result in "returned_partially"
+
+The following will be an example on how to set the order delivery state to "shipped" via the `ship` transition:
+
+```php
+$this->stateMachineRegistry->transition(new Transition(
+    OrderDeliveryDefinition::ENTITY_NAME,
+    '16d5bbd005434d71b45f4b18b9932502',
```
Comment (mitelg) [inline guides/plugins/plugins/checkout/order/using-the-state-machine.md:124]: here you use a real ID, in the examples above only a placeholder

File: guides/plugins/plugins/checkout/order/using-the-state-machine.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        'ship',
+        'stateId'
+    ), $context);
+}
+```
+
+So this example is making use of the repository for the `order_delivery` entity. It was injected previous using the
+[Dependency injection container](../../plugin-fundamentals/dependency-injection.md) and its respective ID `order_delivery.repository`.
+
+Then it's creating a new `Criteria` object and adds a filter in order to only search for `order_delivery´ entities, whose
```
Comment (mitelg) [inline guides/plugins/plugins/checkout/order/using-the-state-machine.md:179]:
```suggestion
Then it's creating a new `Criteria` object and adds a filter in order to only search for `order_delivery` entities, whose
```

**Review by mitelg — CHANGES_REQUESTED** [review]
what about this helper class? https://github.com/shopware/platform/blob/trunk/src/Core/Checkout/Order/Aggregate/OrderTransaction/OrderTransactionStateHandler.php
it is totally missing here

File: guides/plugins/plugins/checkout/order/using-the-state-machine.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
+# Using the state machine
+
+## Overview
+
+An order in Shopware consists of three states:
+- The order state itself, such as "Is the order open?", its technical name being `order`
+- The order transaction state, e.g. "Was it paid?", its technical name being `order_transaction`
+- The order delivery state, e.g. "Was it shipped?", its technical name being `order_delivery`
+
+Each of those comes with several possible values, e.g. for the transactions some of those are: "open", "reopen", "cancel", "chargeback" etc.
```
[the remark this answers is not in the document]
Comment (PaddyS) [reply guides/plugins/plugins/checkout/order/using-the-state-machine.md:10]: Found it [here](https://github.com/shopware/platform/blob/trunk/src/Core/System/StateMachine/Aggregation/StateMachineTransition/StateMachineTransitionActions.php) and in my database. 🤔

[the remark this answers is not in the document]
Comment (PaddyS) [reply guides/plugins/plugins/checkout/order/using-the-state-machine.md:10]: Oh, yeah, I kinda confused those while developing as well. There is a difference between the action names (the ones you need to use) and the state names. E.g. "do_pay" is the action, while "in_progress" is the state afterwards.

I tested chargeback once more and it really didn't work, yet it's a valid action due to the database. Weird.
Removed it.

**Comment by PaddyS** [comment]
> what about this helper class? https://github.com/shopware/platform/blob/trunk/src/Core/Checkout/Order/Aggregate/OrderTransaction/OrderTransactionStateHandler.php
> it is totally missing here

@mitelg Yeah, right. But afaik it's only available for the order transaction states and it felt weird to document like "Oh, also we've got a helper for that, but just for the transactions..."

Still think it's a good idea to add it?

**Comment by mitelg** [comment]
yes please 😁👍 
it is even not used in the core 🙈 

one could argue, why is it still there and not deprecated/removed? but still it is used in most payment plugins

**Review by mitelg — APPROVED** [review]

**Pull request merged**